### PR TITLE
[gpu] Add support for NVIDIA cuDNN

### PR DESCRIPTION
**Pull request by bradmiro**
Adds an option for users to install a user-provided version of CUDNN on Ubuntu-based images.

### CODE REVIEW

**Comment by bradmiro** [comment]
@mengdong Would love your feedback here as well, and any suggestions for possibly adding Debian functionality :)

**Comment by bradmiro** [comment]
Ack indentation issue, will fix shortly.

**Comment by medb** [comment]
`dataproc-1.3-debian10-tests` fails, probably we need to exclude it from this test case?

**Comment by bradmiro** [comment]
> `dataproc-1.3-debian10-tests` fails, probably we need to exclude it from this test case?

Fixed.

File: integration_tests/dataproc_test_case.py [hunk 1]
```diff
@@ -199,6 +199,11 @@ def tearDown(self):
     def getClusterName(self):
         return self.name
 
+    @staticmethod
+    def getBaseOS():
+        # Get the base operating system: '1.5-debian10' -> 'debian' 
+        return re.search("[a-z]+", FLAGS.image_version).group()
```
Comment (medb) [inline integration_tests/dataproc_test_case.py:205]: What it will return for `preview-debian10` version?

File: gpu/test_gpu.py [hunk 2]
```diff
@@ -16,6 +18,10 @@ def verify_instance_gpu_agent(self, name):
     self.assert_instance_command(
         name, "systemctl status gpu-utilization-agent.service")
 
+  def verify_instance_cudnn(self, name):
+    self.assert_instance_command(
+        name, "sudo find /usr -iname libcudnn.so | grep .")
```
Comment (medb) [inline gpu/test_gpu.py:23]: What does `grep .` mean?

File: gpu/test_gpu.py [hunk 3]
```diff
@@ -112,6 +118,32 @@ def test_install_gpu_cuda_nvidia(self, configuration, machine_suffixes,
       self.verify_instance("{}-{}".format(self.getClusterName(),
                                           machine_suffix))
 
+  @parameterized.parameters(
+      ("STANDARD", ["m", "w-0", "w-1"], GPU_V100, GPU_V100, "NVIDIA",
+      "8.0.3.33"),
```
Comment (medb) [inline gpu/test_gpu.py:123]: Seems like this can fit on previous line.

File: gpu/install_gpu_driver.sh [hunk 4]
```diff
@@ -105,6 +108,18 @@ function install_nvidia_nccl() {
     "apt-get install -y --allow-unauthenticated libnccl2=${nccl_version} libnccl-dev=${nccl_version}"
 }
 
+function install_nvidia_cudnn() {
+  if [[ ${OS_NAME} == ubuntu ]]; then 
+    local major_version=${CUDNN_VERSION%%.*}
+
+    apt-get install --no-install-recommends libcudnn${major_version}=${CUDNN_VERSION}-1+cuda${CUDA_VERSION}
+    apt-get install --no-install-recommends libcudnn${major_version}-dev=${CUDNN_VERSION}-1+cuda${CUDA_VERSION}
```
Comment (medb) [inline gpu/install_gpu_driver.sh:116]: Run install with retries and merge in one command:
```bash
local -r cudnn_pkg_version="${CUDNN_VERSION}-1+cuda${CUDA_VERSION}"
execute_with_retries \
    "apt-get install --no-install-recommends libcudnn${major_version}=${cudnn_pkg_version} libcudnn${major_version}-dev=${cudnn_pkg_version}"
```

**Review by medb — CHANGES_REQUESTED** [review]

File: gpu/test_gpu.py (hunk 2 above)
Comment (bradmiro) [reply gpu/test_gpu.py:23]: `grep .` is checking if there's any output from `find`, as `find` always returns exit code zero.

File: integration_tests/dataproc_test_case.py (hunk 1 above)
Comment (bradmiro) [reply integration_tests/dataproc_test_case.py]: Oops, fixed...

File: gpu/test_gpu.py (hunk 2 above)
Comment (medb) [reply gpu/test_gpu.py:25]: Maybe something like this will be more explicit:
```
[[ -n $(sudo find /usr -iname libcudnn.so | wc -l) ]]
```

**Review by medb — CHANGES_REQUESTED** [review]

File: integration_tests/dataproc_test_case.py [hunk 5]
```diff
@@ -199,6 +199,14 @@ def tearDown(self):
     def getClusterName(self):
         return self.name
 
+    @staticmethod
+    def getBaseOS():
+        # Get the base operating system: '1.5-debian10' -> 'debian'
+        # Dataproc default "preview" version is Debian, so "debian" 
+        # is returned if just "preview" is provided for the image version.
+        return "debian" if FLAGS.image_version == "preview" else \
```
Comment (medb) [inline integration_tests/dataproc_test_case.py:207]: Maybe let's make it more generic for all images, i.e. `1.3`, `1.4`, `1.5`?

File: integration_tests/dataproc_test_case.py [hunk 6]
```diff
@@ -199,6 +199,14 @@ def tearDown(self):
...
+            re.search("-[a-z]+", FLAGS.image_version).group()
```
Comment (medb) [inline integration_tests/dataproc_test_case.py:208]: What it will return for `2.0.0-RC22-debian10` version?

**Review by medb — CHANGES_REQUESTED** [review]

File: gpu/test_gpu.py (hunk 2 above)
Comment (bradmiro) [reply gpu/test_gpu.py]: Ack, fixed (with a slight tweak)

File: integration_tests/dataproc_test_case.py (hunk 6 above)
Comment (bradmiro) [reply integration_tests/dataproc_test_case.py]: I think I've covered all edge-cases now, but please let me know if I haven't.

File: integration_tests/dataproc_test_case.py (hunk 5 above)
Comment (bradmiro) [reply integration_tests/dataproc_test_case.py]: Ack, fixed

**Comment by medb** [comment]
Seems like 1.3 tests are failing.

File: gpu/test_gpu.py [hunk 7]
```diff
@@ -70,6 +75,10 @@ def test_install_gpu_without_agent(self, configuration, machine_suffixes,
   def test_install_gpu_with_agent(self, configuration, machine_suffixes,
                                   master_accelerator, worker_accelerator,
                                   driver_provider):
+    # GPU agent supported on Dataproc 1.4+
+    if self.getImageVersion() < pkg_resources.parse_version("1.4"):
+      return
```
Comment (medb) [inline gpu/test_gpu.py:80]: Use skipTest instead (see other tests for an example)

**Review by medb — CHANGES_REQUESTED** [review]

File: gpu/test_gpu.py (hunk 7 above)
Comment (bradmiro) [reply gpu/test_gpu.py]: ack, fixed

File: gpu/test_gpu.py [hunk 8]
```diff
@@ -19,7 +19,7 @@ def verify_instance_gpu_agent(self, name):
 
   def verify_instance_cudnn(self, name):
     self.assert_instance_command(
-        name, "/usr/sbin/ldconfig -p | grep libcudnn" )
+        name, "/sbin/ldconfig -p | grep libcudnn" )
```
Comment (medb) [inline gpu/test_gpu.py:22]: I think `sudo ldconfig -p` will be more universal?

Comment (medb) [inline gpu/test_gpu.py:22]: Does it fail if `grep` output is empty?

**Review by medb — CHANGES_REQUESTED** [review]

File: gpu/test_gpu.py (hunk 8 above)
Comment (bradmiro) [reply gpu/test_gpu.py:22]: Ack

Comment (bradmiro) [reply gpu/test_gpu.py:22]: Yes.

File: gpu/README.md [hunk 9]
```diff
@@ -103,6 +103,32 @@ In current Dataproc set up, we enable GPU resource isolation by initialization
 script without NVIDIA Docker, you can find more information at
 [NVIDIA Spark RAPIDS getting started guide](https://nvidia.github.io/spark-rapids/).
 
+#### cuDNN
+
+You can also install [cuDNN](https://developer.nvidia.com/CUDNN) on your
+cluster. cuDNN is used as a backend for Deep Learning frameworks, such as
+TensorFlow. This feature is currently only supported on Ubuntu-based Dataproc
```
Comment (medb) [inline gpu/README.md:110]: This is not true anymore? I.e. it supported on Debian too?

File: gpu/README.md [hunk 10]
```diff
@@ -125,6 +151,10 @@ script without NVIDIA Docker, you can find more information at
 -   `cuda-version: 10.1|10.2|<VERSION>` - this is an optional parameter for
     customizing NVIDIA-provided CUDA version. Default is `10.2`.
 
+-   `cudnn-version: <VERSION>` - this is an optional parameter for installing
+    [NVIDIA cuDNN](https://developer.nvidia.com/CUDNN) version `x.x.x.x` on
+    Ubuntu-based Dataproc images. There is no default value.
```
Comment (medb) [inline gpu/README.md:156]: Add instructions for Debian images too?

File: gpu/test_gpu.py [hunk 11]
```diff
@@ -112,6 +121,27 @@ def test_install_gpu_cuda_nvidia(self, configuration, machine_suffixes,
...
 
+  @parameterized.parameters(
+      ("STANDARD", ["m", "w-0", "w-1"], GPU_V100, GPU_V100, "NVIDIA", "8.0.5.39"),
+  )
+  def test_install_gpu_with_cudnn(self, configuration, machine_suffixes,
+                                  master_accelerator, worker_accelerator,
+                                  driver_provider, cudnn_version):
+    metadata = "cudnn-version={}".format(cudnn_version)
+    if driver_provider is not None:
+        metadata += ",gpu-driver-provider={}".format(driver_provider)
```
Comment (medb) [inline gpu/test_gpu.py:132]: Fix indentation: 2 spaces instead of 4

**Review by medb — CHANGES_REQUESTED** [review]

File: gpu/README.md (hunk 10 above)
Comment (bradmiro) [reply gpu/README.md]: Fixed (removed Ubuntu-only message)

File: gpu/test_gpu.py (hunk 11 above)
Comment (bradmiro) [reply gpu/test_gpu.py]: Fixed

File: gpu/README.md (hunk 9 above)
Comment (bradmiro) [reply gpu/README.md]: Fixed

**Review by medb — APPROVED** [review]

**Pull request merged**